torch/distributions/transforms.py

```diff
@@ -0,0 +1,342 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all, lazy_property
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'BoltzmannTransform',
+    'ExpTransform',
+    'InverseTransform',
+    'LowerCholeskyTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for invertable transformations with computable log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Caching is useful for tranforms whose inverses are either expensive or
+    numerically unstable. Note that care must be taken with memoized values
+    since the autograd graph may be reversed. For example while the following
+    works with or without caching::
+
+        y = t(x)
+        t.log_abs_det_jacobian(x, y).backward()  # x will receive gradients.
+
+    However the following will error when caching due to dependency reversal::
+
+        y = t(x)
+        z = t.inv(y)
+        grad(z.sum(), [y])  # error because z is x
+
+    Derived classes should implement one or both of :meth:`_call` or
+    :meth:`_inverse`. Derived classes that set `bijective=True` should also
+    implement :meth:`log_abs_det_jacobian`.
+    """
+    bijective = False
+
+    def __init__(self, cache_size=0):
+        if cache_size != 0:
+            if cache_size == 1:
+                self._cached_x_y = None, None
+                self.__call__ = self._cached_call
+                self.inverse = self._cached_inverse
+            else:
+                raise NotImplementedError('cache_size must be 0 or 1')
+
+    @lazy_property
+    def inv(self):
+        """
+        Returns the inverse :class:`Transform` of this transform.
+        """
+        return InverseTransform(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __ne__(self, other):
+        # Necessary for Python2
+        return not self.__eq__(other)
+
+    def __call__(self, x):
+        """
+        Computes the transform `x => y`.
+        """
+        return self._call(x)
+
+    def inverse(self, y):

```

Comment:
is there any reason to keep this method when you can just do `dist.inv(x)`? They should be equivalent

Reply:
Good point. I was able to make this private as `._inv_call()`, but it is still needed to implement `.inv.__call__()`. We want a single object to manage both the `.__call__()` cache and the `.inv.__call__()` cache so that calling one populates the other. Thus `_Inverse` delegates caching to its `.inv` transform.